src/main/tutorials/calculate-metrics/jqassistant/ood-metrics.adoc

```diff
@@ -0,0 +1,315 @@
+RETURN p
+----
+
+[[ood-metrics:AbstractnessTests]]
+[role=group,includesConcepts="ood-metrics:AbstractnessPackageA,ood-metrics:AbstractnessPackageB,ood-metrics:AbstractnessPackageC,ood-metrics:AbstractnessPackageD,ood-metrics:AbstractnessPackageE"]
+
+=== The main sequence
+
+The main sequence is the relationship between abstractness and instability.
+A package is totally abstract or totally instable at the best.
+The packages with high abstractness and high instability are in the zone of uselessness.
+Classes in this zone are maximally abstract and have no dependents.
+The packages with low abstractness and low instability are in the zone of pain.
+Classes in this zone are highly stable and concrete.
+This classes are difficult to change because they are stable, but they cannot be extended because they aren't abstract.
+
+.OO-Metrics Distance
+image::OO-Metrics_Distancequarter.png[OO-Metrics Distance quarter]
+
+Ideally is the distance of a package as low as possible.
+
+==== D (distance)
+
+D = ABS(A + I - 1) / SQRT(2) +
+This is the distance like geometrically defined.
+This metric has the range [0, ~0.707].
+
+[[ood-metrics:Distance]]
+.Computes the distance for all packages
+[source,cypher,role=concept,requiresConcepts="ood-metrics:Abstractness,ood-metrics:Instability"]
+----
+MATCH (p:Package)
+SET p.distance = abs(p.abstractness + p.instability -1) / sqrt(2)
+RETURN p
+----
+
+==== D' (normalized distance)
+
+D' = ABS( A + I - 1) +
+More common is the normalized distance.
+This metric has the range [0, 1].
+
+[[ood-metrics:NormalizedDistance]]
+.Computes the normalized distance for all packages
+[source,cypher,role=concept,requiresConcepts="ood-metrics:Abstractness,ood-metrics:Instability"]
+----
+MATCH (p:Package)
+SET p.normalizedDistance = abs(p.abstractness + p.instability -1)
+RETURN p
+----
+
+==== Example
+
+.OO-Metrics Distance
+image::OO-Metrics_Distance.png[OO-Metrics Distance]
+
+==== Test cases
+
+[[ood-metrics:NormalizedDistancePackageA]]
+.Normalized distance Package a
+[source,cypher,role=concept,requiresConcepts="ood-metrics:NormalizedDistance",severity=major]
+----
+MATCH (p:Package {fqn:"your.project.oodmetrics.a", normalizedDistance:0.25})
+RETURN p
+----
+
+[[ood-metrics:NormalizedDistancePackageB]]
+.Normalized distance Package b
+[source,cypher,role=concept,requiresConcepts="ood-metrics:NormalizedDistance",severity=major]
+----
+MATCH (p:Package {fqn:"your.project.oodmetrics.b", normalizedDistance:0})
+RETURN p
+----
+
+[[ood-metrics:NormalizedDistancePackageC]]
+.Normalized distance Package c
+[source,cypher,role=concept,requiresConcepts="ood-metrics:NormalizedDistance",severity=major]
+----
+MATCH (p:Package {fqn:"your.project.oodmetrics.c", normalizedDistance:0.25})
+RETURN p
+----
+
+[[ood-metrics:NormalizedDistancePackageD]]
+.Normalized distance Package D
+[source,cypher,role=concept,requiresConcepts="ood-metrics:NormalizedDistance",severity=major]
```

Comment:
Attributes like `role` or `requiresConcepts` not visible in the rendered tutorial. Users not so familiar with jQAssistant will not be able to understand it is not a simple query but a concept which depends on another concept. So the dependencies between the various concepts are not comprehensible.